docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -1,95 +1,64 @@
 
-You can try it out by running the example under
-[`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
+If we run this, you'll notice that verfication succeeds.
 
-```
+```bash
 cargo kani --harness safe_update
 ```
 
-## Unsafe nondeterministic variables
-
-Kani also includes an **unsafe** method to generate unconstrained nondeterministic variables which do not take type invariants into consideration.
-As with any unsafe method in Rust, users have to make sure the right guardrails are put in place to avoid undesirable behavior.
-
-That said, there may be cases where you want to verify your code taking into consideration that some inputs may contain invalid data.
-
-Let's see what happens if we modify our verification harness to use the unsafe method `kani::any_raw()` to generate the updated value.
-
-```rust
-{{#include tutorial/arbitrary-variables/src/inventory.rs:unsafe_update}}
-```
-
-We commented out the assertion that the underlying `u32` variable cannot be `0`, since this no longer holds.
-The verification will now fail showing that the `inventory.get(&id).unwrap()` method call can panic.
+`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
+For `NonZeroU32`, this means we never return a `0` value.
+The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.
 
-This is an interesting issue that emerges from how `rustc` optimizes the memory layout of `Option<NonZeroU32>`.
-The compiler is able to represent `Option<NonZeroU32>` using `32` bits by using the value `0` to represent `None`.
+## Other kinds of nondeterministic variables
 
-You can try it out by running the example under [`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+There's nothing special about `kani::any()` except that Kani ships with it only implemented for types where we can guarantee safety.
+Notably, however, there is no implementation for `Vec<T>`, for example.
 
-```
-cargo kani --harness unsafe_update
-```
+The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, as we saw in the last chapter.
+There are no arguments provided to `kani::any()` to indicate an upper bound.
 
-## Safe nondeterministic variables for custom types
+Likewise, `kani::any()` is implemented for `[T; N]` (if implemeneted for `T`).
+But, it is _not_ implemented for `&[T]`.
+Again, there's no upper bound information we can make use of.
 
-Now you would like to add a new structure to your library that allow users to represent a review rating, which can go from 0 to 5 stars.
-Let's say you add the following implementation:
+This does not mean you cannot have a nondeterministic vector.
+You just have to construct it.
```

Comment:
But how? I think this is missing. We could document the Arbitrary type here. That is a very safe trait.